src/frontend/fcitx4frontend/fcitx4frontend.cpp

```diff
@@ -0,0 +1,357 @@
+    std::string name_;
+};
+
+std::tuple<int, bool, uint32_t, uint32_t, uint32_t, uint32_t>
+Fcitx4InputMethod::createICv3(const std::string &appname, int pid) {
+    auto sender = currentMessage()->sender();
+    int icid = module_->nextIcIdx();
+    auto *ic = new Fcitx4InputContext(icid, instance_->inputContextManager(),
+                                      this, sender, appname);
+    bus_->addObjectVTable(ic->path().path(), FCITX_INPUTCONTEXT_DBUS_INTERFACE,
+                          *ic);
+
+    return std::make_tuple(icid, false, 0, 0, 0, 0);
+}
+
+char *setStrWithLen(char *res, const char *str, size_t len) {
+    if (res) {
+        res = static_cast<char *>(realloc(res, len + 1));
+    } else {
+        res = static_cast<char *>(malloc(len + 1));
+    }
+    memcpy(res, str, len);
+    res[len] = '\0';
+    return res;
+}
+
+int getDisplayNumber() {
+    const char *display = getenv("DISPLAY");
+    if (!display)
+        return 0;
+    size_t len;
+    const char *p = display + strcspn(display, ":");
+    if (*p != ':')
+        return 0;
+    p++;
+    len = strcspn(p, ".");
+    char *str_disp_num = setStrWithLen(nullptr, p, len);
+    int displayNumber = atoi(str_disp_num);
+    free(str_disp_num);
+    return displayNumber;
+}
+
+std::string readFileContent(const std::string &file) {
+    std::ifstream fin(file, std::ios::binary | std::ios::in);
+    std::vector<char> buffer;
+    constexpr auto chunkSize = 4096;
+    do {
+        auto curSize = buffer.size();
+        buffer.resize(curSize + chunkSize);
+        if (!fin.read(buffer.data() + curSize, chunkSize)) {
+            buffer.resize(curSize + fin.gcount());
+            break;
+        }
+    } while (0);
+    std::string str{buffer.begin(), buffer.end()};
+    return stringutils::trim(str);
+}
+
+std::string getLocalMachineId() {
+    auto content = readFileContent("/var/lib/dbus/machine-id");
+    if (content.empty()) {
+        content = readFileContent("/etc/machine-id");
+    }
+
+    if (content.empty()) {
+        content = "machine-id";
+    }
+
+    return content;
+}
+
+Fcitx4FrontendModule::Fcitx4FrontendModule(Instance *instance)
+    : instance_(instance),
+      portalBus_(std::make_unique<dbus::Bus>(dbus::BusType::Session)),
+      Fcitx4InputMethod_(
+          std::make_unique<Fcitx4InputMethod>(this, bus(), "/inputmethod")),
+      Fcitx4InputMethodCompatible_(std::make_unique<Fcitx4InputMethod>(
+          this, portalBus_.get(), "/inputmethod")) {
+    Flags<dbus::RequestNameFlag> requestFlag =
+        dbus::RequestNameFlag::ReplaceExisting;
+    this->bus()->requestName(FCITX_DBUS_SERVICE, requestFlag);
+    auto display = getDisplayNumber();
+    auto dbusServiceName =
+        stringutils::concat(FCITX_DBUS_SERVICE, "-", display);
+    this->bus()->requestName(dbusServiceName, requestFlag);
+
+    auto localMachineId = getLocalMachineId();
+    auto path = stringutils::joinPath(
```

Comment:
use StandardPath::safeSave and check for failure.

You may simply throw exception on failure.